sos/report/plugins/loki.py

```diff
@@ -0,0 +1,190 @@
+            f"--data-urlencode 'query={{{label}=~\".+\"}}' "
+            f"--data-urlencode 'start={start_formatted}' "
+            f"--data-urlencode 'end={end_formatted}' "
+            f"--data-urlencode 'limit={Loki.LOKI_QUERY_LIMIT}' "
+        )
+        return command
+
+    def get_logs(
+            self,
+            endpoint,
+            label,
+            start: Optional[datetime],
+            end: Optional[datetime]):
+        output = self.exec_cmd(self.query_command(endpoint, label, start, end))
+        logs = json.loads(output["output"])
+        return logs
+
+    def get_earliest_log_timestamp(self, logs):
+        log_streams = logs["data"]["result"]
+        # use now as a comparison
+        earliest_log = int(datetime.now().timestamp()*1_000_000_000)
+        for stream in log_streams:
+            for log in stream["values"]:
+                timestamp = int(log[0])
+                earliest_log = min(earliest_log, timestamp)
+        return earliest_log
+
+    def get_logs_for_label(self, endpoint, label, paginate):
+        logs = self.get_logs(endpoint, label, None, None)
+        with self.collection_file(f'{label}.log') as logfile:
+            logfile.write(json.dumps(logs, indent=2))
+        if paginate:
+            earliest_log = self.get_earliest_log_timestamp(logs)
+            previous_earliest_log = int(
+                datetime.now().timestamp()*1_000_000_000
+            )
+            iterations_count = 0
+            while iterations_count < Loki.MAX_PAGINATION_ITERATIONS and \
+                    earliest_log < previous_earliest_log:
+                new_logs = self.get_logs(
+                    endpoint,
+                    label,
+                    None,
+                    datetime.fromtimestamp(earliest_log / 1_000_000_000)
+                )
+                previous_earliest_log = earliest_log
+                earliest_log = \
+                    self.get_earliest_log_timestamp(new_logs)
+                with self.collection_file(f'{label}.log.{iterations_count}') \
+                        as logfile:
+                    logfile.write(json.dumps(new_logs, indent=2))
+                # exit at most after 100 pages to avoid infinite loops
+                iterations_count += 1
+
+    def setup(self):
+        els_config_file = self.path_join(
+            "/etc/loki/*.yaml"
+        )
+        self.add_copy_spec(els_config_file)
+
+        # charms using cos-coordinated-workers have their config elsewhere
+        coordinated_workers_config_file = self.path_join(
+            "/etc/worker/*.yaml"
+        )
+        self.add_copy_spec(coordinated_workers_config_file)
+
+        self.add_copy_spec("/var/log/loki/*")
+        self.add_cmd_output(["pebble logs loki -n 10000"])
+
+        if self.get_option("collect-logs") is True:
+            endpoint = self.get_option("endpoint") or "http://localhost:3100"
+            labels = []
+            if self.get_option("labels"):
```

Comment:
You could use a walrus operatot

```python
            if labels_option := self.get_option("labels"):
```

Then you can ditch the `labels_option = self.get_option("labels")` the line below